patchwerk-sim.py

```diff
@@ -74,42 +74,95 @@ def __str__(self):
             name = "Healer #{} Heal".format(self._entity)
         return "[Time {}] {}".format(time, name)
 
+class Tank:
+    def __init__(self, name, max_health=11000, dodge_parry=0.3, mitigation=0.7):
+        self.name = name
+        self.max_health = max_health
+        self.dodge_parry = dodge_parry
+        self.mitigation = mitigation
+        self.current_health = self.max_health
+
+    def reset(self):
+        self.current_health = self.max_health
+
+    # returns True if tank dies, False otherwise
+    def get_smashed(self, dmg):
+        # check for misses
+        if random.random() < self.dodge_parry:
+            # print("Hateful Strike MISSES {}".format(self.name))
+            return False
+
+        # print("Hateful Strike hits {} ({} hp) for {} dmg".format(self.name, self.current_health, dmg))
+        self.current_health -= dmg
+        if self.current_health <= 0:
+            # print("{} has DIED! ({} Overkill)".format(self.name, -self.current_health))
+            return True
+        return False
+
+    # returns a tuple with total raw healing and overhealing
+    def get_healed(self, heal_qty):
+        # print("{} ({} hp) is healed for {}".format(self.name, self.current_health, heal_qty))
+        self.current_health += heal_qty
+        overhealing = 0
+        if self.current_health > self.max_health:
+            overhealing = self.current_health - self.max_health
+            self.current_health = self.max_health
+
+        return (heal_qty, overhealing)
+
+    def __str__(self):
+        return self.name
+
 # updated hateful strike to hit every 1.2s instead of random number from 1.2 to 2s
 def get_timetonext_hateful():
     return 1.2
-    # seconds_later = (random.random() * (0.8499) + 1.2)
-#     return round(seconds_later, 1)
 
-def get_hateful_target(tanks_health):
-    return tanks_health.index(max(tanks_health))
+# returns a tuple of (tank_to_be_hit_next, tank_index)
+def get_hateful_target(tanks):
+    highest_health_tank_index = -1
+    highest_health_tank = -999
+
+    for index, tank in enumerate(tanks):
```

Comment:
the optimization you mentioned to me, i think it's not worth doing here since it's only 3 OTs.. premature optimization root of all evil etc. if it were a much bigger number than 3 then a max-heap would do perfectly